ai/core/src/unitycatalog/ai/core/oss.py

```diff
@@ -0,0 +1,647 @@
+        function_head: The function signature as a string.
+    """
+
+    function_def: str
+    function_head: str
+
+
+def dynamically_construct_python_function(function_info: FunctionInfo) -> FunctionDefinition:
+    """
+    Construct a Python function from the given FunctionInfo.
+
+    Args:
+        function_info: The FunctionInfo object containing the function metadata.
+
+    Returns:
+        A FunctionDefinition named tuple containing the function definition and function head.
+    """
+
+    param_names = []
+    if function_info.input_params and function_info.input_params.parameters:
+        param_names = [param.name for param in function_info.input_params.parameters]
+    function_head = f"{function_info.name}({', '.join(param_names)})"
+    func_def = f"def {function_head}:\n"
+    if function_info.routine_body == "EXTERNAL":
+        for line in function_info.routine_definition.split("\n"):
+            func_def += f"    {line}\n"
+    else:
+        raise NotImplementedError(f"routine_body {function_info.routine_body} not supported")
+
+    return FunctionDefinition(function_def=func_def, function_head=function_head)
+
+
+def validate_input_parameter(
+    parameter: Union[FunctionParameterInfo, Dict[str, str]],
+) -> FunctionParameterInfo:
+    """
+    Validate the input parameter and convert it to FunctionParameterInfo.
+
+    Args:
+        parameter: The input parameter to validate. If it is a dict, it will be converted to FunctionParameterInfo.
+
+    Returns:
+        The validated FunctionParameterInfo.
+    """
+
+    if isinstance(parameter, dict):
+        parameter = FunctionParameterInfo(**parameter)
+    elif not isinstance(parameter, FunctionParameterInfo):
+        raise TypeError(
+            f"Input parameter should be either a dict or an instance of "
+            f"FunctionParameterInfo, but got {type(parameter)}"
+        )
+
+    required_fields = ["name", "type_name", "type_text", "type_json"]
+
+    if missing_fields := [
+        field for field in required_fields if getattr(parameter, field, None) is None
+    ]:
+        raise ValueError(
+            f"Missing required fields in input parameter '{getattr(parameter, 'name', '')}': {missing_fields}."
+        )
+
+    if isinstance(type_name := parameter.type_name, Enum):
+        type_name = type_name.value
+    elif not isinstance(type_name, str):
+        raise ValueError(
+            f"Invalid type_name '{type_name}' in input parameter '{parameter.name}'. Must be a string or Enum instance."
+        )
+
+    complex_type_names = ["ARRAY", "MAP", "STRUCT", "DECIMAL", "INTERVAL"]
+    if any(type_name.startswith(complex_type_name) for complex_type_name in complex_type_names):
+        parameter.type_name = next(
+            complex_type_name
+            for complex_type_name in complex_type_names
+            if type_name.startswith(complex_type_name)
+        )
+        return parameter
```

Comment:
```suggestion
    for complex_type_name in complex_type_names:
        if type_name.startswith(complex_type_name):
            parameter.type_name = complex_type_name
            return parameter
```
Could we use a simple for loop?

Reply:
lol good call :) much simpler with a basic loop